File: java/src/main/java/com/cisco/pxgrid/samples/ise/http/SessionQueryByIP.java
package com.cisco.pxgrid.samples.ise.http;

import java.io.IOException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.cisco.pxgrid.model.AccountState;
import com.cisco.pxgrid.model.Service;

/**
 * Demonstrates how to query a session using IP address
 */
public class SessionQueryByIP {
	private static Logger logger = LoggerFactory.getLogger(SessionQueryByIP.class);

	private static void query(SampleConfiguration config, String ip) throws IOException {
		PxgridControl pxgrid = new PxgridControl(config);
		Service[] services = pxgrid.lookupService("com.cisco.ise.session");
		if (services == null || services.length == 0) {
			System.out.println("Service unavailabe");
			return;
		}
		Service service = services[0];
		String url = service.getProperties().get("restBaseUrl") + "/getSessionByIpAddress";
		logger.info("url={}", url);
		String secret = pxgrid.getAccessSecret(service.getNodeName());
		String postData = "{\"ipAddress\":\"" + ip + "\"}";
		SampleHelper.postStringAndPrint(url, config.getNodeName(), secret, config.getSSLContext().getSocketFactory(), postData);
	}
	

	public static void main(String [] args) throws Exception {
		SampleConfiguration config = new SampleConfiguration();
		PxgridControl pxgrid = new PxgridControl(config);
		
		while (pxgrid.accountActivate() != AccountState.ENABLED)
			Thread.sleep(60000);
		logger.info("pxGrid controller version=" + pxgrid.getControllerVersion());

		while (true) {
			String ip = SampleHelper.prompt("IP address (or <enter> to disconnect): ");
			if (ip == null)	break;
			query(config, ip);
		}
	}
}

File: python/session_query_by_ip.py
from pxgrid import PxgridControl
from config import Config
import urllib.request
import base64
import time
import logging
import json

logger = logging.getLogger(__name__)


def query(config, secret, url, payload):
    handler = urllib.request.HTTPSHandler(context=config.ssl_context)
    opener = urllib.request.build_opener(handler)
    rest_request = urllib.request.Request(url=url, data=str.encode(payload))
    rest_request.add_header('Content-Type', 'application/json')
    rest_request.add_header('Accept', 'application/json')
    b64 = base64.b64encode((config.node_name + ':' + secret).encode()).decode()
    rest_request.add_header('Authorization', 'Basic ' + b64)
    rest_response = opener.open(rest_request)
    return rest_response.read().decode()


if __name__ == '__main__':
    config = Config()

    #
    # verbose logging if configured
    #
    if config.verbose:
        handler = logging.StreamHandler()
        handler.setFormatter(logging.Formatter('%(asctime)s:%(name)s:%(levelname)s:%(message)s'))
        logger.addHandler(handler)
        logger.setLevel(logging.DEBUG)

        # and set for stomp and ws_stomp modules also
        for stomp_mod in ['stomp', 'ws_stomp', 'pxgrid']:
            s_logger = logging.getLogger(stomp_mod)
            handler.setFormatter(logging.Formatter('%(asctime)s:%(name)s:%(levelname)s:%(message)s'))
            s_logger.addHandler(handler)
            s_logger.setLevel(logging.DEBUG)

    pxgrid = PxgridControl(config=config)

    while pxgrid.account_activate()['accountState'] != 'ENABLED':
        time.sleep(60)

    # lookup for session service
    service_lookup_response = pxgrid.service_lookup('com.cisco.ise.session')
    service = service_lookup_response['services'][0]
    node_name = service['nodeName']
    url = service['properties']['restBaseUrl'] + '/getSessionByIpAddress'

    secret = pxgrid.get_access_secret(node_name)['secret']

    if not config.ip:
        ip = input('Enter IP address: ')
    else:
        ip = config.ip
    resp = query(config, secret, url, '{ "ipAddress": "%s" }' % ip)
    print(json.dumps(json.loads(resp), indent=2, sort_keys=True))
